h/auth/services.py

```diff
@@ -0,0 +1,114 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+import datetime
+
+from pyramid_authsanity import interfaces
+import sqlalchemy as sa
+from zope import interface
+from zope.sqlalchemy import mark_changed
+
+from h.models import AuthTicket
+from h.auth.util import principals_for_user
+
+TICKET_TTL = datetime.timedelta(days=7)
+
+
+class AuthTicketNotLoadedError(Exception):
+    pass
+
+
+@interface.implementer(interfaces.IAuthService)
+class AuthTicketService(object):
+    def __init__(self, session, user_service):
+        self.session = session
+        self.usersvc = user_service
+
+        self._userid = None
+        self._user = None
+
+    def userid(self):
+        """
+        Return current userid, or None.
+
+        Raises ``AuthTicketNotLoadedError`` when auth ticket has not been
+        loaded yet, which signals the auth policy to call ``verify_ticket``.
+        """
+
+        if self._userid is None:
+            raise AuthTicketNotLoadedError('auth ticket is not loaded yet')
+
+        if self._user is None:
+            self._user = self.usersvc.fetch(self._userid)
+
+        return self._user.userid
+
+    def groups(self):
+        """Returns security principals of the logged-in user."""
+
```

Comment:
That's fine because `pyramid_authsanity` only ever calls `authsvc.groups()` after first calling `authsvc.userid()` (and, if necessary, `authsvc.verify_ticket(..., ...)`).